brave/brave-ios
# Fix #5490: Refactor TransactionView to improve performance

## Summary of Changes
- Extracted the displayed data for `TransactionView` into a `TransactionSummary` model. Previously we were using the `WeiFormatter` on every SwiftUI layout pass, this model allows us to generate the formatted values once and store in a model instead of on each layout pass.
- Created a `TransactionParser` to help generate this `TransactionSummary` model given a transaction and necessary values (the current network, user accounts, user visible assets, all tokens, and asset price ratios).
- Replaced `TransactionView` with a `TransactionSummaryView` and updated `AccountActivityStore` / `AccountActivityView` and `AssetDetailStore` / `AssetDetailView` to use this new `TransactionSummary` model and `TransactionSummaryView`.
- Added the new `ethSwap` so we display titles for swap transactions again.

This pull request fixes #5490

## Submitter Checklist:


## Test Plan:
If you have lots of transactions on one account
1. Open & unlock Wallet
2. Tap Accounts
3. Tap account with lots of transactions
4. Scroll the list of transactions up/down quickly, should now see smooth scrolling without micro-stutters

If you have lots of transactions on one token (can be across all accounts):
1. Open  & unlock Wallet
2. Tap token with lots of transactions in visible assets list
3. Scroll the list of transactions up/down quickly, should now see smooth scrolling without micro-stutters

Developers: Although most of the logic was copied when moving to the `TransactionSummary` / `TransactionParser` builder, I screenshotted both views against `development` and this branch to ensure the transactions display the same. Currently on `development` you will have to add `.ethSwap` to the case statements [here (Transactions/TransactionView.swift#L95)](https://github.com/brave/brave-ios/blob/development/BraveWallet/Crypto/Transactions/TransactionView.swift#L95) and [here (AssetDetailStore.swift#L185)](https://github.com/brave/brave-ios/blob/development/BraveWallet/Crypto/Stores/AssetDetailStore.swift#L185) to match up (otherwise swaps will be excluded as we don't correctly handle the new TxType).


## Screenshots:

Updated video from my XS (original stuttering video in Issue #5490):

https://user-images.githubusercontent.com/5314553/173398547-cee10b3c-7593-4a2a-968c-16277990a35f.MP4


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
